src/utils.js

```diff
@@ -0,0 +1,170 @@
+
+const isDigit = (ch) => (ch >= "0" && ch <= "9");
+
+const isLiteral = (ch) => isAlpha(ch) || isDigit(ch) || ' :/'.includes(ch);
+
+const isSet = (term) => isOpenSet(term[0]) && isCloseSet(term.slice(-1));
+
+const isAlternate = (term) => isOpenAlternate(term[0]) && isCLoseAlternate(term.slice(-1));
+
+const isUnit = (term) => isLiteral(term[0]) || isDot(term[0]) || isSet(term) || isEscapeSequence(term);
+
+const isOpenSet = (char) => char === '[';
+
+const isCloseSet = (char) => char === ']';
+
+const splitSet = (set_head) => set_head.slice(1, -1).split('');
+
+const splitAlternate = (alternate) => alternate.slice(1, -1).split('|');
+
+function removeSlashWrapper(expr) {
+  if (expr[0] === '/' && expr.slice(-1) === '/') {
+    return expr.slice(1, -1);
+  } else {
+    throw new Error('Incorrect RegExp. The RegExp must be wrapped in slashes');
+  }
+}
+
+function duplicate(str, n) {
+  let res = '';
+  for (let i = 0; i < n; i++) {
+    res += str;
+  }
+  return res;
+}
+
+function splitExpr(expr) {
+  let head = '';
+  let operator = '';
+  let rest = '';
+  let last_expr_pos = 0;
+
+  if (isOpenSet(expr[0])) {
+    last_expr_pos = expr.indexOf(']') + 1;
+    head = expr.slice(0, last_expr_pos);
+  } else if (isOpenAlternate(expr[0])) {
+    last_expr_pos = expr.indexOf(')') + 1;
+    head = expr.slice(0, last_expr_pos);
+  } else if (isEscapeSequence(expr)) {
+    last_expr_pos += 2;
+    head = expr.slice(0, 2);
+  } else {
+    last_expr_pos = 1;
+    head = expr[0];
+  }
+
+  if (last_expr_pos < expr.length && isOperator(expr[last_expr_pos])) {
+    operator = expr[last_expr_pos];
+    last_expr_pos++;
+  }
+
+  rest = expr.slice(last_expr_pos);
+
+  return [head, operator, rest];
+}
+
+function doesUnitMatch(expr, str) { 
+  const [head, operator, rest] = splitExpr(expr);
+
+  if (str.length === 0) return false;
+  if (isLiteral(head)) return expr[0] === str[0];
+  if (isDot(head)) return true;
+  if (isEscapeSequence(head)) {
+    if (head === '\\a') return isAlpha(str[0]);
+    if (head === '\\d') return isDigit(str[0]);
+    return false;
+  }
+  if (isSet(head)) {
+    const set_terms = splitSet(head);
+    return set_terms.includes(str[0]);
+  }
+  return false;
+}
+
+function matchMultiple(expr, str, match_length, min_length = 0, max_length = -1) {
+  const [head, operator, rest] = splitExpr(expr);
+  let submatch_length = -1;
+
+  while (max_length === -1 || (submatch_length < max_length)) {
+    const [subexpr_matched] = matchExpr(
+      duplicate(head, submatch_length + 1), 
+      str, 
+      match_length
+    );
+    if (!subexpr_matched) break;
+    submatch_length++;
+  }
+
+  while (submatch_length >= min_length) {
+    const [matched, new_match_length] = matchExpr(
+        duplicate(head, submatch_length) + rest, 
+        str, 
+        match_length
+    );
+    if (matched) return [matched, new_match_length];
+    submatch_length--;
+  }
+
+  return [false, 0];
```

Comment:
Here is the same as was mentioned above. Now it's a bit confusing. Make the function return object with proper fields names